src/adapter/src/util.rs

```diff
@@ -448,31 +448,41 @@ pub(crate) fn viewable_variables<'a>(
         })
 }
 
-/// Verify that the row datums match the expected desc.
-pub fn verify_datum_desc(desc: &RelationDesc, rows: &[Row]) -> Result<(), AdapterError> {
+/// Verify that the rows in [`RowIterator`] match the expected [`RelationDesc`].
+pub fn verify_datum_desc(
+    desc: &RelationDesc,
+    rows: &mut dyn RowIterator,
+) -> Result<(), AdapterError> {
     // Verify the first row is of the expected type. This is often good enough to
-    // find problems. Notably it failed to find #6304 when "FETCH 2" was used in a
-    // test, instead we had to use "FETCH 1" twice.
-    if let [row, ..] = rows {
-        let datums = row.unpack();
-        let col_types = &desc.typ().column_types;
-        if datums.len() != col_types.len() {
+    // find problems.
+    //
+    // Notably it failed to find #6304 when "FETCH 2" was used in a test, instead
+    // we had to use "FETCH 1" twice.
+
+    let Some(row) = rows.peek() else {
+        return Ok(());
+    };
+
+    let datums = row.unpack();
+    let col_types = &desc.typ().column_types;
+    if datums.len() != col_types.len() {
+        let msg = format!(
+            "internal error: row descriptor has {} columns but row has {} columns",
+            col_types.len(),
+            datums.len(),
+        );
+        return Err(AdapterError::Internal(msg));
+    }
```

Comment:
Not covered by any test?

Reply:
Makes sense! This code is only hit if we have an internal bug where the returned Rows do not match the expected RelationDesc